sandbox: request remote execution output as a directory digest

abderrahim opened the pull request:
This gives us an output that is easier to manipulate given we're using buildbox-casd, and allows to avoid a whole class of bugs such as #1842 and #1961

The old code for handling the output as a Tree is kept for now as the remote execution spec recommends using it as a fallback

abderrahim committed:
sandbox: request remote execution output as a directory digest

This gives us an output that is easier to manipulate given we're using
buildbox-casd, and allows to avoid a whole class of bugs such as #1842
and #1961

The old code for handling the output as a Tree is kept for now as the
remote execution spec recommends using it as a fallback

juergbi commented on src/buildstream/sandbox/_sandboxremote.py:
```diff
@@ -294,12 +294,20 @@ def _execute_action(self, action, flags):
+
+            root_directory_digest = output_directory.root_directory_digest
+            if root_directory_digest and root_directory_digest.hash:
+                cascache._fetch_directory(casremote, root_directory_digest)
```
Nit, `_fetch_directory()` should be marked as public (internal) by removing the leading underscore as we're calling it from outside `cascache.py`.

abderrahim committed:
cas: make CASCache._fetch_directory() internal

It is used by the artifact/source caches and will be used by the
remote execution sandbox
